homeassistant/components/zwave_js/__init__.py

```diff
@@ -182,50 +186,77 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     # Set up websocket API
     async_register_api(hass)
 
-    # Create a task to allow the config entry to be unloaded before the driver is ready.
-    # Unloading the config entry is needed if the client listen task errors.
-    start_client_task = hass.async_create_task(start_client(hass, entry, client))
-    hass.data[DOMAIN].setdefault(entry.entry_id, {})[
-        DATA_START_CLIENT_TASK
-    ] = start_client_task
-
-    return True
-
-
-async def start_client(
-    hass: HomeAssistant, entry: ConfigEntry, client: ZwaveClient
-) -> None:
-    """Start listening with the client."""
-    entry_hass_data: dict = hass.data[DOMAIN].setdefault(entry.entry_id, {})
-    entry_hass_data[DATA_CLIENT] = client
-    driver_events = entry_hass_data[DATA_DRIVER_EVENTS] = DriverEvents(hass, entry)
+    driver_ready = asyncio.Event()
+    listen_task = hass.async_create_background_task(
+        client_listen(hass, entry, client, driver_ready),
+        f"{DOMAIN}_{entry.title}_client_listen",
+    )
 
     async def handle_ha_shutdown(event: Event) -> None:
         """Handle HA shutdown."""
-        await disconnect_client(hass, entry)
+        await client.disconnect()
 
-    listen_task = asyncio.create_task(
-        client_listen(hass, entry, client, driver_events.ready)
-    )
-    entry_hass_data[DATA_CLIENT_LISTEN_TASK] = listen_task
     entry.async_on_unload(
         hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, handle_ha_shutdown)
     )
 
+    driver_ready_task = hass.async_create_task(driver_ready.wait())
     try:
-        await driver_events.ready.wait()
-    except asyncio.CancelledError:
-        LOGGER.debug("Cancelling start client")
-        return
+        async with asyncio.timeout(LISTEN_READY_TIMEOUT):
+            done, _ = await asyncio.wait(
```

Comment:
Maybe. We'd be able to remove two lines. That's the only reason, I think.

Reply:
Its a big confusing to figure out whats going to happen, and which task gets cancelled first here since `asyncio.timeout` will cancel the task, and `asyncio.wait` will not cancel the task. Personally I try not to mix them as it gives me a headache tracking down issues later.

Reply:
Ok, sounds good.